File: app/src/main/java/com/luigi573/mira/di/AppComponent.java
package com.luigi573.mira.di;

import com.luigi573.mira.MiraApp;
import com.luigi573.mira.data.source.StatisticDataSourceModule;

import javax.inject.Singleton;

import dagger.BindsInstance;
import dagger.Component;
import dagger.android.AndroidInjectionModule;
import dagger.android.AndroidInjector;
import dagger.android.support.AndroidSupportInjectionModule;

@Singleton
@Component(modules = {AndroidInjectionModule.class
        , AndroidSupportInjectionModule.class
        , ActivityBindingModule.class
        , AppModule.class
        , StatisticDataSourceModule.class})
public interface AppComponent extends AndroidInjector<MiraApp> {


    @Component.Builder
    interface Builder {
        @BindsInstance
        AppComponent.Builder application(MiraApp application);

        AppComponent build();
    }

}

File: app/src/main/java/com/luigi573/mira/di/AppComponent.kt
package com.luigi573.mira.di

import android.content.Context
import com.luigi573.mira.MiraApp
import com.luigi573.mira.data.source.StatisticDataSourceModule
import com.luigi573.mira.statistics.StatisticsModule
import dagger.BindsInstance
import dagger.Component
import dagger.android.AndroidInjectionModule
import dagger.android.AndroidInjector
import dagger.android.support.AndroidSupportInjectionModule
import javax.inject.Singleton

@Singleton
@Component(
    modules = [
        AndroidInjectionModule::class,
        AndroidSupportInjectionModule::class,
        AppModule::class,
        StatisticsModule::class,
        StatisticDataSourceModule::class
    ])
interface AppComponent : AndroidInjector<MiraApp> {
    @Component.Factory
    interface Factory {
        fun create(@BindsInstance applicationContext: Context): AppComponent
    }
}